File: LeetCode/13. Roman to Integer/Solution.cc
#include <cstdlib>
#include <string>

using std::string;

class Solution {
public:
  int romanToInt(const std::string& s) {
    int result = 0;

    size_t i = 0;
    while (i < s.size()) {
      char c = s[i];
      int value = map(c);

      if (i + 1 < s.size() && value < map(s[i + 1])) {
        result += map(s[i + 1]) - value;
        i += 2;
      } else {
        result += value;
        i++;
      }
    }

    return result;
  }

  /**
   * Symbol Value I 1 V 5 X 10 L 50 C 100 D 500 M 1000
   */
  int map(char c) {
    switch (c) {
    case 'I':
      return 1;
    case 'V':
      return 5;
    case 'X':
      return 10;
    case 'L':
      return 50;
    case 'C':
      return 100;
    case 'D':
      return 500;
    case 'M':
      return 1000;
    default:
      std::abort();
    }
  }
};

// test

#include <iostream>

void test(const std::string &input, int expected) {
  std::cout << input << " ->" << Solution().romanToInt(input)
            << " == " << expected << std::endl;
}

int main() {
  test("MCMXCIV", 1994);
  test("LVIII", 58);
  test("III", 3);
}

File: LeetCode/13. Roman to Integer/Solution.kt
package problem13_1

class Solution {
  /**
   * Symbol Value I 1 V 5 X 10 L 50 C 100 D 500 M 1000
   */
  val table = mapOf(
    'I' to 1,
    'V' to 5,
    'X' to 10,
    'L' to 50,
    'C' to 100,
    'D' to 500,
    'M' to 1000
  )

  fun romanToInt(s: String): Int {
    var result = 0

    var i = 0
    while (i < s.length) {
      val char = s[i]
      val value = table[char]!!

      if (i + 1 < s.length && value < table[s[i + 1]]!!) {
        result += table[s[i + 1]]!! - value
        i += 2

      } else {
        result += value
        i++
      }
    }

    return result
  }
}

// test

fun test(
  // input
  s: String,
  // expected
  expected: Int,
) {

  println("${Solution().romanToInt(s)} == ${expected}")
}

fun main() {
  test("III", 3)
  test("LVIII", 58)
  test("MCMXCIV", 1994)
}